pkg/controller/servicebindingrequest/mapper.go

```diff
@@ -12,7 +21,97 @@ var (
 
 // sbrRequestMapper is the handler.Mapper interface implementation. It should influence the
 // enqueue process considering the resources informed.
-type sbrRequestMapper struct{}
+type sbrRequestMapper struct {
+	client     dynamic.Interface
+	restMapper meta.RESTMapper
+}
+
+var serviceBindingRequestGVK = v1alpha1.SchemeGroupVersion.WithKind("ServiceBindingRequest")
+var secretGVK = corev1.SchemeGroupVersion.WithKind("Secret")
+
+// isServiceBindingRequest checks whether the given obj is a Service Binding Request through GVK
+// comparison.
+func isServiceBindingRequest(obj runtime.Object) bool {
+	return obj.GetObjectKind().GroupVersionKind() == serviceBindingRequestGVK
+}
+
+// isSecret checks whether the given obj is a Secret through GVK comparison.
+func isSecret(obj runtime.Object) bool {
+	return obj.GetObjectKind().GroupVersionKind() == secretGVK
+}
+
+// isSBRService checks whether the given obj is a service in given sbr.
+func isSBRService(sbr *v1alpha1.ServiceBindingRequest, obj runtime.Object) bool {
+	services := extractServiceSelectors(sbr)
+	for _, svc := range services {
+		svcGVK := schema.GroupVersionKind{Group: svc.Group, Version: svc.Version, Kind: svc.Kind}
+		if obj.GetObjectKind().GroupVersionKind() == svcGVK {
+			return true
+		}
+	}
+	return false
+}
+
+// isApplication checks whether the given obj is an application in given sbr.
+func isApplication(
+	restMapper meta.RESTMapper,
+	sbr *v1alpha1.ServiceBindingRequest,
+	obj runtime.Object,
+) (bool, error) {
+	appGVR := schema.GroupVersionResource{
+		Group:    sbr.Spec.ApplicationSelector.Group,
+		Version:  sbr.Spec.ApplicationSelector.Version,
+		Resource: sbr.Spec.ApplicationSelector.Resource,
+	}
+	appGVK, err := restMapper.KindFor(appGVR)
+	if err != nil {
+		return false, err
+	}
+
+	isEqual := obj.GetObjectKind().GroupVersionKind() == appGVK
+
+	return isEqual, nil
+}
+
+// isSecretOwnedBySBR checks whether the given obj is a secret owned by the given sbr.
+func isSecretOwnedBySBR(obj metav1.Object, sbr *v1alpha1.ServiceBindingRequest) bool {
```

Comment:
The correct way to check if a secret is owned by an SBR would be through owner reference.

Reply:
"Owning" here is a loose term;  the relationship between a `corev1.Secret` and a `ServiceBindingRequest` can happen in two distinct ways: one is as intermediate/intermediary secret through SBR's `.status.secret` and another is through services relationship.

In other words, a SBR owns the intermediate secret when the secret lives in the same namespace as the SBR and its name is present in `.status.secret`.

Reply:
@isuttonl I don't understand this comment :slightly_smiling_face: 

Isn't it a hack to determine the relationship between an intermediate secret and an SBR based on intermediate secret's name when we can do that with ownerReference?

Just trying to understand,
This intermediate secret was created by the Service Binding Operator.

It makes sense to check relationship between SBR - application based on name present in spec.Application.ResourceRef because SBR doesn't OWN it.

But in the case of intermediate secret, There is a more concrete way of checking the relationship with SBR.

Let me know if my understanding is not right. :) 

You mentioned "owning" here is a loose term. Why is that so?

Reply:
There's a couple of ownership concepts here:

1. Ownership from Kubernetes' garbage collector point-of-view, where the SBR is present in the Secret's `OwnerReferences` field; since it is a slice and lives in the resource's metadata and can be modified by the user, I believe it should be used only to coordinate resource deletion (blocking or cascading, for example).
2. Ownership from a SBO business logic point-of-view, represented by the `.status.secret` field in SBR. In other words the, intermediary/intermediate secret name is stored there and does not interfere with the garbage collector. 

I meant "loose" because the function is responsible for identifying 2) not 1) as you've previously thought.

Reply:
Another point is that mapper is responsible for identifying the relationship of an arbitrary resource to a slice `ServiceBindingRequest` namespace/name pairs, and currently there are 3 distinct places this is encoded in the `ServiceBindingRequest`:

1. Applications
2. Services
3. `.status.secret`

Reply:
Okay. Thanks for the explaination.